src/libraries/Microsoft.Extensions.DependencyInjection/src/ScopePool.cs

```diff
@@ -5,40 +5,54 @@
 using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Diagnostics;
-using System.Threading;
 using Microsoft.Extensions.DependencyInjection.ServiceLookup;
 
 namespace Microsoft.Extensions.DependencyInjection
 {
     internal class ScopePool
     {
-        // Modest number to re-use. We only really care about reuse for short lived scopes
-        private const int MaxQueueSize = 128;
+        // Stash the scope state on a thread static
+        [ThreadStatic]
+        private static State t_tlsScopeState;
+        private volatile int _maxScopeSize;
 
-        private int _count;
-        private readonly ConcurrentQueue<State> _queue = new();
+        // Roughly ~3.6KB
+        private const int MaxResolvedServicesCount = 100;
+
+        // Below default ~85K LOH threshold, avoid large temporary garbage
+        private const int MaxCapacity = 2000;
 
         public State Rent()
         {
-            if (_queue.TryDequeue(out State state))
+            // Try to get a scope from TLS
+            State state = t_tlsScopeState;
+
+            if (state != null)
             {
-                Interlocked.Decrement(ref _count);
-                return state;
+                // Clear the state
+                t_tlsScopeState = null;
             }
-            return new State(this);
+
+            return state ?? new State(this, _maxScopeSize);
         }
 
         public bool Return(State state)
         {
-            if (Interlocked.Increment(ref _count) > MaxQueueSize)
+            var size = state.Size;
+
+            if (size <= MaxResolvedServicesCount)
             {
-                Interlocked.Decrement(ref _count);
-                return false;
+                // Stash the state back in TLS
+                state.Clear();
+                t_tlsScopeState = state;
```

Comment:
Rent and Return won't always run on the same thread, right? You could have two scopes in progress and after async/await both return their scopes on the same thread, with the second returned scope overwriting the first.

Reply:
Yep, but I don't see why that matters.

Reply:
It just means that the hit rate of the pool will vary based on how perfectly distributed the requests are. If you are lucky, close to everything will hit the pool. If you are very unlucky, close to nothing will hit the pool.

Reply:
I'm a glass always full type of person.